appinventor/components/src/com/google/appinventor/components/runtime/ChatBot.java

```diff
@@ -634,6 +673,42 @@ private synchronized void ensureSslSockFactory() {
     }
   }
 
+  /*
+   * Parse response from the Chatbot. Extract the image data from the
+   * returned protocol buffer, if provided. Call either GotResponse,
+   * or GotResponseWithImage depending on whether or not an image was
+   * returned.
+   *
+   * @param parsed protocol buffer returned from proxy
+   * @returns Void
+   */
+  private void parseResponse(ChatBotToken.response response) {
+    String responseText = response.getAnswer();
+    byte[] responseImage = response.getOutputimage().toByteArray();
+    if (DEBUG) {
+      Log.d(LOG_TAG, "parseResponse: responseImage = " + responseImage);
+    }
+    if (responseImage != null && responseImage.length > 0) {
```

Comment:
Note that if the bot doesn't return an image, it requires the user to have to implement both the GotResponse and GotResponseWithImage event handlers. We should track the `doImage` state from `performRequest` and only call the event handler that responds to the block that initiated the request.